File: server/agent/src/llm/gemini.js
const { GoogleGenerativeAI } = require('@google/generative-ai');
require('dotenv').config();

const MODEL_NAME = 'gemini-2.5-flash';
let genAI = null;
let model = null;

function init() {
    if (model) return;
    const apiKey = process.env.GEMINI_API_KEY;
    if (!apiKey) throw new Error('GEMINI_API_KEY not set in .env');
    genAI = new GoogleGenerativeAI(apiKey);
    model = genAI.getGenerativeModel({ model: MODEL_NAME });
}

function toSystemInstruction(text) {
    return { role: 'user', parts: [{ text }] };
}

async function generateResponse(systemPrompt, userMessage, history = []) {
    init();

    const chatModel = genAI.getGenerativeModel({
        model: MODEL_NAME,
        systemInstruction: toSystemInstruction(systemPrompt),
        tools: [{ googleSearch: {} }]
    });

    const chat = chatModel.startChat({
        history: history.map(msg => ({
            role: msg.role === 'assistant' ? 'model' : msg.role,
            parts: [{ text: msg.content }]
        }))
    });

    let attempts = 0;
    const maxAttempts = 3;

    while (attempts < maxAttempts) {
        try {
            const result = await chat.sendMessage(userMessage);
            const text = result.response.text();

            let groundingMetadata = null;
            try {
                const candidate = result.response.candidates?.[0];
                const gm = candidate?.groundingMetadata;
                if (gm && gm.groundingChunks && gm.groundingChunks.length > 0) {
                    groundingMetadata = {
                        searchQueries: gm.webSearchQueries || [],
                        sources: gm.groundingChunks
                            .filter(c => c.web)
                            .map(c => ({ url: c.web.uri, title: c.web.title }))
                    };
                }
            } catch (e) { }

            return { success: true, text, source: 'gemini', groundingMetadata };
        } catch (error) {
            attempts++;
            if (attempts >= maxAttempts) {
                return { success: false, error: error.message, source: 'gemini' };
            }

            await new Promise(r => setTimeout(r, 1000 * Math.pow(2, attempts)));
        }
    }
}

async function generateJSON(systemPrompt, userMessage) {
    init();

    const jsonModel = genAI.getGenerativeModel({
        model: MODEL_NAME,
        systemInstruction: toSystemInstruction(systemPrompt),
        generationConfig: { responseMimeType: 'application/json' }
    });

    let attempts = 0;
    const maxAttempts = 3;

    while (attempts < maxAttempts) {
        try {
            const result = await jsonModel.generateContent(userMessage);
            const text = result.response.text();
            const parsed = JSON.parse(text);
            return { success: true, data: parsed, raw: text, source: 'gemini' };
        } catch (error) {
            attempts++;
            if (attempts >= maxAttempts) {
                return { success: false, error: error.message, source: 'gemini' };
            }
            await new Promise(r => setTimeout(r, 1000 * Math.pow(2, attempts)));
        }
    }
}

/**
 * Generate a response with function calling tools (for ReAct pattern).
 * Returns either a text response or a function call request.
 * 
 * @param {string} systemPrompt — system instruction for the model
 * @param {string} userMessage — the user's message
 * @param {Object[]} toolDeclarations — array of { name, description, parameters } tool definitions
 * @param {Object[]} history — conversation history
 * @returns {Promise<Object>} — { success, type: 'text'|'functionCall', text?, functionCall?: { name, args } }
 */
async function generateWithTools(systemPrompt, userMessage, toolDeclarations = [], history = []) {
    init();

    const modelConfig = {
        model: MODEL_NAME,
        systemInstruction: toSystemInstruction(systemPrompt)
    };

    // Add function declarations as tools if provided
    if (toolDeclarations.length > 0) {
        modelConfig.tools = [{
            functionDeclarations: toolDeclarations
        }];
    }

    const toolModel = genAI.getGenerativeModel(modelConfig);

    const chat = toolModel.startChat({
        history: history.map(msg => ({
            role: msg.role === 'assistant' ? 'model' : msg.role,
            parts: [{ text: msg.content }]
        }))
    });

    let attempts = 0;
    const maxAttempts = 3;

    while (attempts < maxAttempts) {
        try {
            const result = await chat.sendMessage(userMessage);
            const response = result.response;
            const candidate = response.candidates?.[0];

            if (!candidate) {
                return { success: false, error: 'No candidate in response', source: 'gemini' };
            }

            // Check if the model wants to call a function
            const parts = candidate.content?.parts || [];
            const functionCallPart = parts.find(p => p.functionCall);

            if (functionCallPart) {
                const fc = functionCallPart.functionCall;
                console.log(`[Gemini] Function call requested: ${fc.name}`);
                return {
                    success: true,
                    type: 'functionCall',
                    functionCall: {
                        name: fc.name,
                        args: fc.args || {}
                    },
                    source: 'gemini'
                };
            }

            // Otherwise, it's a text response (final answer)
            const textPart = parts.find(p => p.text);
            const text = textPart ? textPart.text : response.text();

            return {
                success: true,
                type: 'text',
                text: text,
                source: 'gemini'
            };
        } catch (error) {
            attempts++;
            console.warn(`[Gemini] generateWithTools attempt ${attempts}/${maxAttempts} failed:`, error.message);
            if (attempts >= maxAttempts) {
                return { success: false, error: error.message, source: 'gemini' };
            }
            await new Promise(r => setTimeout(r, 1000 * Math.pow(2, attempts)));
        }
    }
}

async function healthCheck() {
    try {
        init();
        const result = await model.generateContent('Reply with OK');
        return result.response.text().length > 0;
    } catch {
        return false;
    }
}

module.exports = { generateResponse, generateJSON, generateWithTools, healthCheck };

File: server_py/agent/llm/gemini.py
"""
Gemini LLM — Google Generative AI SDK integration (google-genai).
"""

import asyncio
import json
from typing import Optional

from google import genai
from google.genai import types

from config import get_settings

MODEL_NAME = "gemini-2.5-flash"
_client: Optional[genai.Client] = None


def get_client() -> genai.Client:
    """Get or create the Gemini client singleton."""
    global _client
    if _client is not None:
        return _client

    settings = get_settings()
    api_key = settings.GEMINI_API_KEY
    if not api_key:
        raise RuntimeError("GEMINI_API_KEY not set in .env")

    _client = genai.Client(api_key=api_key)
    return _client


def _build_history(history: list[dict]) -> list[types.Content]:
    """Convert conversation history to Gemini format."""
    result = []
    for msg in history:
        role = "model" if msg.get("role") == "assistant" else msg.get("role", "user")
        result.append(types.Content(
            role=role,
            parts=[types.Part.from_text(text=msg.get("content", ""))],
        ))
    return result


async def generate_response(
    system_prompt: str,
    user_message: str,
    history: list[dict] = None,
) -> dict:
    """Generate a text response with Google Search grounding.

    Returns:
        { success, text, source, groundingMetadata? }
    """
    history = history or []
    client = get_client()

    attempts = 0
    max_attempts = 3

    while attempts < max_attempts:
        try:
            contents = _build_history(history)
            contents.append(types.Content(
                role="user",
                parts=[types.Part.from_text(text=user_message)],
            ))

            response = await asyncio.to_thread(
                client.models.generate_content,
                model=MODEL_NAME,
                contents=contents,
                config=types.GenerateContentConfig(
                    system_instruction=system_prompt,
                    tools=[types.Tool(google_search=types.GoogleSearch())],
                ),
            )

            text = response.text or ""

            # Extract grounding metadata
            grounding_metadata = None
            try:
                candidate = response.candidates[0] if response.candidates else None
                if candidate and hasattr(candidate, "grounding_metadata"):
                    gm = candidate.grounding_metadata
                    if gm and hasattr(gm, "grounding_chunks") and gm.grounding_chunks:
                        grounding_metadata = {
                            "searchQueries": list(gm.web_search_queries or []),
                            "sources": [
                                {"url": c.web.uri, "title": c.web.title}
                                for c in gm.grounding_chunks
                                if hasattr(c, "web") and c.web
                            ],
                        }
            except Exception:
                pass

            return {"success": True, "text": text, "source": "gemini", "groundingMetadata": grounding_metadata}

        except Exception as error:
            attempts += 1
            if attempts >= max_attempts:
                return {"success": False, "error": str(error), "source": "gemini"}
            await asyncio.sleep(2 ** attempts)


async def generate_json(system_prompt: str, user_message: str) -> dict:
    """Generate a JSON response.

    Returns:
        { success, data, raw, source }
    """
    client = get_client()

    attempts = 0
    max_attempts = 3

    while attempts < max_attempts:
        try:
            response = await asyncio.to_thread(
                client.models.generate_content,
                model=MODEL_NAME,
                contents=user_message,
                config=types.GenerateContentConfig(
                    system_instruction=system_prompt,
                    response_mime_type="application/json",
                ),
            )

            text = response.text or ""
            parsed = json.loads(text)
            return {"success": True, "data": parsed, "raw": text, "source": "gemini"}

        except Exception as error:
            attempts += 1
            if attempts >= max_attempts:
                return {"success": False, "error": str(error), "source": "gemini"}
            await asyncio.sleep(2 ** attempts)


async def generate_with_tools(
    system_prompt: str,
    user_message: str,
    tool_declarations: list[dict] = None,
    history: list[dict] = None,
) -> dict:
    """Generate a response with function calling tools (for ReAct pattern).

    Returns:
        { success, type: 'text'|'functionCall', text?, functionCall?: { name, args } }
    """
    tool_declarations = tool_declarations or []
    history = history or []
    client = get_client()

    # Build tool definitions for the SDK
    tools = []
    if tool_declarations:
        function_declarations = []
        for td in tool_declarations:
            function_declarations.append(types.FunctionDeclaration(
                name=td["name"],
                description=td["description"],
                parameters=td.get("parameters"),
            ))
        tools = [types.Tool(function_declarations=function_declarations)]

    attempts = 0
    max_attempts = 3

    while attempts < max_attempts:
        try:
            contents = _build_history(history)
            contents.append(types.Content(
                role="user",
                parts=[types.Part.from_text(text=user_message)],
            ))

            response = await asyncio.to_thread(
                client.models.generate_content,
                model=MODEL_NAME,
                contents=contents,
                config=types.GenerateContentConfig(
                    system_instruction=system_prompt,
                    tools=tools if tools else None,
                ),
            )

            if not response.candidates:
                return {"success": False, "error": "No candidate in response", "source": "gemini"}
            candidate = response.candidates[0]
            if not candidate or not candidate.content:
                return {"success": False, "error": "Empty candidate content", "source": "gemini"}

            parts = candidate.content.parts or []

            # Check for function call
            for part in parts:
                if hasattr(part, "function_call") and part.function_call:
                    fc = part.function_call
                    print(f"[Gemini] Function call requested: {fc.name}")
                    return {
                        "success": True,
                        "type": "functionCall",
                        "functionCall": {
                            "name": fc.name,
                            "args": dict(fc.args) if fc.args else {},
                        },
                        "source": "gemini",
                    }

            # Text response (final answer)
            text = ""
            for part in parts:
                if hasattr(part, "text") and part.text:
                    text = part.text
                    break

            if not text:
                text = response.text or ""

            return {
                "success": True,
                "type": "text",
                "text": text,
                "source": "gemini",
            }

        except Exception as error:
            attempts += 1
            print(f"[Gemini] generateWithTools attempt {attempts}/{max_attempts} failed: {error}")
            if attempts >= max_attempts:
                return {"success": False, "error": str(error), "source": "gemini"}
            await asyncio.sleep(2 ** attempts)


async def health_check() -> bool:
    """Check if Gemini is available."""
    try:
        client = get_client()
        response = await asyncio.to_thread(
            client.models.generate_content,
            model=MODEL_NAME,
            contents="Reply with OK",
        )
        return bool(response.text and len(response.text) > 0)
    except Exception:
        return False
